### DISCUSSION TITLE
SEARCH_FILTER syntax for multiple hostnames

### DISCUSSION THREAD

**Discussion by s-m-p** [CrowdStrike/falconpy#817] [Q&A] [answered]
Apologies for such a newbie question...I'm experimenting with falconpy for the first time and I'm not very experienced with Python. I'm starting from the Quick Start example and struggling with the SEARCH_FILTER syntax. I simply want to pass a list of hostnames to the filter. I've tried all the permutations I could think of based on an API call that that I was able to run successfully from Postman. Maybe I'm supposed to edit the 'query_devices_by_filter' function, but I'm not sure. I studied the FQL doc too, but still can't figure it out. Can someone give me a brief example to clarify what SEARCH_FILTER needs to look like in order to pass multiple hostnames? 

This particular attempt produced the error "No hosts found matching that hostname within your Falcon tenant."
```python
SEARCH_FILTER = 'host1,host2'

hosts_search_result = hosts.query_devices_by_filter(filter=f"hostname:*'*{SEARCH_FILTER}*'")

```

**Comment by jshcodes** [comment, marked as answer]
Hi @s-m-p -

You have a couple of options here depending on how you need to architect your search.  A stemmed search that would match several hosts of similar hostnames is the standard and what you see above, but there are additional filters available to you as well.

You could pass a list to the filter, please note this would only return exact matches: 

```python
hosts_search_result = hosts.query_devices_by_filter(filter=f"hostname:['HOSTNAME1','HOSTNAME2','ETC']")
```

You can also build a complex filter (up to 20 properties) similar to:

> OR matches (note the comma)
```python
hosts_search_result = hosts.query_devices_by_filter(filter=f"hostname:'HOSTNAME1',hostname:'HOSTNAME2_OR',hostname:'HOSTNAME3_OR'")
```

> AND matches (plus sign for this variation)
```python
hosts_search_result = hosts.query_devices_by_filter(filter=f"hostname:'HOSTNAME1'+hostname:'HOSTNAME2_AND'+hostname:'HOSTNAME3_AND'")
```

You can get more complex leveraging the `(` and `)` parenthesis.

More details on [FQL (Falcon Query Language)](https://www.falconpy.io/Usage/Falcon-Query-Language.html) can be found on [falconpy.io](https://falconpy.io). You're also welcome to post any questions you encounter here. 

Thank you for the question! :smile:

**Reply by s-m-p** [reply]
Thanks for the quick reply and all the examples! I should have noted that I'm looking for an exact match. But if I can define SEARCH_FILTER with as a list, shouldn't this work? I get "[Error 400] Invalid filter expression supplied":
```python
    SEARCH_FILTER = ['host1','host2']

    hosts_search_result = hosts.query_devices_by_filter(filter=f"hostname:*'*{SEARCH_FILTER}*'")
```

**Reply by jshcodes** [reply]
Try it like this:

```python
SEARCH_FILTER = ['host1', 'host2']
host_search_result = hosts.query_devices_by_filter(filter=f"hostname:{SEARCH_FILTER}")
```

😄

**Reply by s-m-p** [reply]
That doesn't seem to be working.
![Snap22](https://user-images.githubusercontent.com/38626196/199317808-58a40683-5542-4f92-acac-208d2987b94d.jpg)

**Reply by jshcodes** [reply]
That's strange, your f-string looks like it's formatted properly.  I tried it in my console before posting (and just tried it again with valid and invalid hostnames).

```shell
Python 3.9.9 (main, Nov 21 2021, 03:23:42)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> from falconpy import Hosts
>>> hosts = Hosts(client_id=os.getenv("FALCON_CLIENT_ID"), client_secret=os.getenv("FALCON_CLIENT_SECRET"))
>>> SEARCH_FILTER = ["Host1", "Host2"]
>>> hosts.query_devices_by_filter_scroll(filter=f"hostname:{SEARCH_FILTER}")
```

```json
{
	"status_code": 200,
	"headers": {
		"Server": "nginx",
		"Date": "Tue, 01 Nov 2022 20:12:43 GMT",
		"Content-Type": "application/json",
		"Content-Length": "198",
		"Connection": "keep-alive",
		"Content-Encoding": "gzip",
		"Strict-Transport-Security": "max-age=15724800; includeSubDomains, max-age=31536000; includeSubDomains",
		"X-Cs-Region": "us-1",
		"X-Cs-Traceid": "REDACTED",
		"X-Ratelimit-Limit": "6000",
		"X-Ratelimit-Remaining": "5999"
	},
	"body": {
		"meta": {
			"query_time": 0.004435169,
			"pagination": {
				"offset": 0,
				"limit": 100,
				"total": 0
			},
			"powered_by": "device-api",
			"trace_id": "REDACTED"
		},
		"resources": [],
		"errors": []
	}
}
```

```shell
>>> print(f"hostname:{SEARCH_FILTER}")
```

```shell
hostname:['Host1', 'Host2']
```
Can you add a print right before the call to `query_devices_by_filter_scroll` so we can confirm your filter string matches the format? (redact hostnames like you did above still)  Since it's a list, technically we might want to wrap it in a str() call (but we really shouldn't _have_ to...).

**Reply by s-m-p** [reply]
There must have been a special character in my IDE somewhere because after fiddling around for a while and coming back to it after doing other stuff, I'm getting the results I would expect. I'm well on my way now. Thanks a lot for clarifying!
